core/src/main/java/feign/ResponseInterceptor.java

```diff
@@ -13,26 +13,101 @@
  */
 package feign;
 
+import static feign.FeignException.errorReading;
+import feign.codec.DecodeException;
+import feign.codec.Decoder;
 import java.io.IOException;
+import java.lang.reflect.Type;
 import java.util.function.Function;
 
 /**
- * Zero or One {@code ResponseInterceptor} may be configured for purposes such as verify or modify
- * headers of response, verify the business status of decoded object. Once interceptors are applied,
- * {@link ResponseInterceptor#aroundDecode(Response, Function)} is called around decode method
- * called
+ * Interceptor for purposes such as verify or modify headers of response, verify the business status
+ * of decoded object. Once interceptors are applied,
+ * {@link ResponseInterceptor#intercept(Response, Function)} is called around decode method called
  */
 public interface ResponseInterceptor {
 
-  ResponseInterceptor DEFAULT = InvocationContext::proceed;
-
   /**
-   * Called for response around decode, must either manually invoke
-   * {@link InvocationContext#proceed} or manually create a new response object
+   * Called for response around decode, must either manually invoke {@link Chain#next(Context)} or
+   * manually create a new response object
    *
    * @param invocationContext information surrounding the response been decoded
    * @return decoded response
    */
-  Object aroundDecode(InvocationContext invocationContext) throws IOException;
+  Object intercept(Context context, Chain chain) throws IOException;
+
+  /**
+   * Return a new {@link ResponseInterceptor} that invokes the current interceptor first and then
+   * the one that is passed in.
+   * 
+   * @param nextInterceptor the interceptor to delegate to after the current
+   * @return a new interceptor that chains the two
+   */
+  default ResponseInterceptor andThen(ResponseInterceptor nextInterceptor) {
```

Comment:
This feels an awful log like `Function.andThen` or `Consumer.andThen`.  Do you think there's a way we could use those instead?